src/main/java/org/isf/opd/manager/OpdBrowserManager.java

```diff
@@ -114,6 +121,29 @@ protected void validateOpd(Opd opd, boolean insert) throws OHDataValidationExcep
 				errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.opd.specifyingduplicatediseasesisnotallowed.msg")));
 			}
 		}
+		try {
+			Disease opdDisease;
+			if (disease != null) {
+				opdDisease = diseaseBrowserManager.getOPDDiseaseByCode(disease.getCode());
+				if (opdDisease == null) {
+					errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.opd.specifieddiseaseisnoenabledforopdservice.msg")));
+				}
+			}
+			if (disease2 != null) {
+				opdDisease = diseaseBrowserManager.getOPDDiseaseByCode(disease2.getCode());
+				if (opdDisease == null) {
+					errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.opd.specifieddiseaseisnoenabledforopdservice.msg")));
+				}
+			}
+			if (disease3 != null) {
+				opdDisease = diseaseBrowserManager.getOPDDiseaseByCode(disease3.getCode());
+				if (opdDisease == null) {
+					errors.add(new OHExceptionMessage(MessageBundle.getMessage("angal.opd.specifieddiseaseisnoenabledforopdservice.msg")));
+				}
+			}
```

Comment:
instead to have 3 queries on the DB (`getOPDDiseaseByCode`) wouldn't be better to get the list (`getDiseaseOpd`) and cycle over it?

by the moment the error message is the one prompted to the user, wouldn't be better to specify also which disease the user should change?

Reply:
There are 90 OPD related diseases.  To fetch and iterate over the list seemed to me to be more expensive then a simple DB query that either returns `null` or a single record.

Besides at least in the sample data the vast majority of OPD records only have a single disease listed; so it is a single query that returns a lot less number of records then getting them all.

As for the error message I was following the example of duplicate diseases that didn't specify which ones were the duplicates.

Reply:
> There are 90 OPD related diseases. To fetch and iterate over the list seemed to me to be more expensive then a simple DB query that either returns null or a single record.
>
> Besides at least in the sample data the vast majority of OPD records only have a single disease listed; so it is a single query that returns a lot less number of records then getting them all.

Ok, I agree

> As for the error message I was following the example of duplicate diseases that didn't specify which ones were the duplicates.

Ok, but while the "duplicate" warning is somehow self-explaining and doesn't need to be verbose (I can "see" what I'm sending to the BE), for the diseases' internal properties it is better to explain IMO, because they depends by the specific installation settings provided by the admin and the client could not necessarily acknowledged about it. For this reason I would even roll back the `angal.opd.specifyingduplicatediseasesisnotallowed.fmt.msg`.